src/index.js

```diff
@@ -0,0 +1,979 @@
+    }
+
+    let iWant = this._handleIHave(peer, controlMsg)
+    let iHave = this._handleIWant(peer, controlMsg)
+    let prune = this._handleGraft(peer, controlMsg)
+    this._handlePrune(peer, controlMsg)
+
+    if (!iWant || !iHave || !prune) {
+      return
+    }
+
+    let outRpc = this._rpcWithControl(null, iHave, iWant, null, prune)
+    this._sendRpc(rpc.from, outRpc)
+  }
+
+  /**
+   * Process incoming messages,
+   * emitting locally and forwarding on to relevant floodsub and gossipsub peers
+   */
+  _processRpcMessages (msgs) {
+    msgs.forEach((msg) => {
+      const seqno = utils.msgId(msg.from, msg.seqno)
+      // Have we seen this message before> if so, ignore
+      if (this.seenCache.has(seqno)) {
+        return
+      }
+
+      this.seenCache.put(seqno)
+
+      const topics = msg.topicIDs
+
+      // Emit to self
+      this._emitMessages(topics, [msg])
+
+      // Emit to floodsub peers
+      this.peers.forEach((peer) => {
+        if (peer.info.protocols.has(constants.FloodSubID) &&
+          peer.info.id.toB58String() !== msg.from &&
+          utils.anyMatch(peer.topics, topics) &&
+          peer.isWritable
+        ) {
+          peer.sendMessages(utils.normalizeOutRpcMessages([msg]))
+          this.log('publish msg on topics - floodsub', topics, peer.info.id.toB58String())
+        }
+      })
+
+      // Emit to peers in the mesh
+      topics.forEach((topic) => {
+        if (!this.mesh.has(topic)) {
+          return
+        }
+        this.mesh.get(topic).forEach((peer) => {
+          if (!peer.isWritable || peer.info.id.toB58String() === msg.from) {
+            return
+          }
+          peer.sendMessages(utils.normalizeOutRpcMessages([msg]))
+          this.log('publish msg on topic - meshsub', topic, peer.info.id.toB58String())
+        })
+      })
+    })
+  }
+
+  _emitMessages (topics, messages) {
+    topics.forEach((topic) => {
+      if (this.subscriptions.has(topic)) {
+        messages.forEach((message) => {
+          this.emit(topic, message)
+        })
+      }
+    })
+  }
+
+  /**
+   * Returns a buffer of a RPC message that contains a control message
+   *
+   * @param {Array<rpc.RPC.Message>} msgs
+   * @param {Array<rpc.RPC.ControlIHave>} ihave
+   * @param {Array<rpc.RPC.ControlIWant>} iwant
+   * @param {Array<rpc.RPC.ControlGraft>} graft
+   * @param {Array<rpc.RPC.Prune>} prune
+   *
+   * @returns {rpc.RPC}
+   *
+   */
+  _rpcWithControl (msgs, ihave, iwant, graft, prune) {
+    return {
+      subscriptions: [],
+      msgs: msgs || [],
+      control: {
+        ihave: ihave || [],
+        iwant: iwant || [],
+        graft: graft || [],
+        prune: prune || []
+      }
+    }
+  }
+
+  /**
+   * Handles IHAVE messages
+   *
+   * @param {Peer} peer
+   * @param {rpc.RPC.controlMessage Object} controlRpc
+   *
+   * @returns {rpc.RPC.ControlIWant Object}
+   */
+  _handleIHave (peer, controlRpc) {
+    let iwant = new Set()
+
+    let ihaveMsgs = controlRpc.ihave
```

Comment:
Looking at the protobuf definition, I think `ihave` will never be `null`, only the `RPC.control` parent object.
If `RPC.control` is not `null`, it will have `ihave`, `iwant`, `graft`, and `prune` arrays.

Reply:
The point here is that you don't need to do the next validation if you add this fallback, which IMHO is way easier for readability.

Reply:
Gotcha. Agreed, much more readable assuming its an Array.